src/Microsoft.OData.Client/Materialization/FeedAndEntryMaterializerAdapter.cs

```diff
@@ -296,6 +323,87 @@ private ODataResourceSet ReadFeedCore(bool lazy)
             return result;
         }
 
+        /// <summary>
+        /// Reads an <see cref="ODataDeltaResourceSet"/>.
+        /// </summary>
+        /// <returns>The <see cref="MaterializerDeltaFeed"/> of the read <see cref="ODataDeltaResourceSet"/>.</returns>
+        private MaterializerDeltaFeed ReadDeltaFeedCore()
+        {
+            this.ExpectState(ODataReaderState.DeltaResourceSetStart);
+
+            ODataDeltaResourceSet result = (ODataDeltaResourceSet)this.reader.Item;
+            MaterializerDeltaFeed feed = null;
+
+            if (result == null)
+            {
+                return null;
+            }
+
+            feed = MaterializerDeltaFeed.CreateDeltaFeed(
+                    result,
+                    new List<IMaterializerState>(),
+                    this.materializerContext);
+
+            if (this.items?.Count > 0)
+            {
+                this.items.Push(feed);
+            }
+            else
+            {
+                IMaterializerState topLevelItem = feed;
+                this.items.Push(topLevelItem);
+            }
+
+            do
+            {
+                this.AssertRead();
+
+                switch (this.reader.State)
+                {
+                    case ODataReaderState.ResourceStart:
+                        MaterializerEntry entry = this.ReadEntryCore();
+                        feed.AddEntry(entry);
+                        break;
+                    case ODataReaderState.DeletedResourceStart:
+                        MaterializerDeletedEntry deletedMaterializerEntry = this.ReadDeletedResource();
+                        feed.AddEntry(deletedMaterializerEntry);
+                        break;
+                    case ODataReaderState.DeletedResourceEnd:
+                        break;

```

Comment:
Why are we not "popping" a deleted resource from `this.items`?

Reply:
Maybe remove the `break` to allow fall-through...